Vim: highlight #function macro, improve string interpolation highlighting

mossheim opened the pull request:
This PR changes the vim plugin's syntax highlighting so that #function is also recognized as a preprocessor macro, and so that these macros are highlighted even when they appear in string interpolation. This matches the behavior of Xcode's highlighting and the highlighting provided by GitHub:

```swift
let mango: [Double] = []
#function
#line
"\(#function), \(data)"
"\(#line)"
```

Before this change:

![pasted image at 2017_08_08 04_43 pm](https://user-images.githubusercontent.com/15369640/29096648-cca08922-7c63-11e7-9e6e-4a3acb834c9d.png)

After:

![pasted image at 2017_08_08 04_44 pm](https://user-images.githubusercontent.com/15369640/29096649-cef299cc-7c63-11e7-9e2c-c884aa75d8dc.png)

---

To my knowledge, this doesn't resolve any bugs in the bug tracker.

It's not clear to me if this counts as a trivial check-in, since it doesn't impact any code for Swift or simulators. If I should run the CI tests please let me know.

mossheim committed:
vim-syntax: highlight #function preproc macro

mossheim committed:
vim-syntax: highlight preproc macros in string interpolation

huonw commented on utils/vim/syntax/swift.vim:
```diff
@@ -140,7 +140,7 @@ syn match swiftTypeDeclaration skipwhite nextgroup=swiftType
 
 syn region swiftString start=/"/ skip=/\\\\\|\\"/ end=/"/ contains=swiftInterpolation
-syn region swiftInterpolation start=/\\(/ end=/)/ contained
+syn region swiftInterpolation start=/\\(/ end=/)/ contained contains=swiftPreproc
```
Should this include all the various expression categories like `swiftCastOp`?

brianlheim replied:
Good point. Really, I'd prefer if nothing but the delimiters were highlighted - this seems to fit more closely with how Xcode is doing it. IMO the `foo` in `"\(foo)"` shouldn't be highlighted. But I don't exactly know how to do that. I think I got close with:

```vim
syn region swiftStringFull start=/"/ skip=/\\\\\|\\"/ end=/"/ contains=TOP,swiftString keepend transparent
syn match swiftString /[")][^"]\{-}\("\|\\(\)/ contained
```

I could certainly add those other categories and keep the existing relationship between `swiftString` and `swiftInterpolation` if you think that's easiest.

huonw commented:
@swift-ci Please smoke test

Even for something like this which is trivial we're required to run tests (minimizing the gray areas/just throwing more CPU time at the problem means minimizing the areas where us pesky humans can make mistakes and accidentally merge a broken thing that isn't as trivial as it looks).

harlanhaskins reviewed (approved):
LGTM as-is, though it may be advantageous to add more entries to `contains` to catch more expressions, like @huonw said.

mossheim committed:
vim syntax: make string interpolation a matchgroup

Now, only the opening \( and ) are highlighted, and everything inside retains its coloring

brianlheim commented:
@swift-ci Please smoke test

@huonw - I think I've got it now. I switched to using a matchgroup for `swiftInterpolation`, and added `contains=TOP` to reset the highlighting context. I think this pretty much matches the way Xcode highlights inside string interpolation, with extra highlighting for the delimiters. Let me know what you think:

![image](https://user-images.githubusercontent.com/15369640/29135892-988f0f0e-7d09-11e7-9262-0440ceb7e806.png)

harlanhaskins commented:
@swift-ci Please smoke test

harlanhaskins reviewed (approved):
LGTM. Thanks!

brianlheim commented:
Thanks @harlanhaskins!
